File: clinical/src/main/java/com/hartwig/actin/clinical/feed/questionnaire/QuestionnaireReader.java
package com.hartwig.actin.clinical.feed.questionnaire;

import java.util.List;
import java.util.Set;
import java.util.StringJoiner;

import com.google.common.collect.Lists;

import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;

final class QuestionnaireReader {

    static final Set<String> TERMS_TO_CLEAN =
            Set.of("{", "}", "\\tab", "\\li0", "\\ri0", "\\sa0", "\\sb0", "\\fi0", "\\ql", "\\par", "\\f2", "\\ltrch");

    private QuestionnaireReader() {
    }

    @NotNull
    public static String[] read(@NotNull QuestionnaireEntry entry, @NotNull List<String> validKeys) {
        return merge(clean(entry.itemAnswerValueValueString()).split("\n"), validKeys);
    }

    @NotNull
    private static String[] merge(@NotNull String[] lines, @NotNull List<String> validKeys) {
        List<String> merged = Lists.newArrayList();

        StringJoiner curLine = newValueStringJoiner();
        for (int i = 0; i < lines.length; i++) {
            curLine.add(lines[i]);
            if (lines[i].isEmpty() || i == lines.length - 1 || isField(lines[i + 1], validKeys) || lines[i + 1].isEmpty()) {
                merged.add(curLine.toString());
                curLine = newValueStringJoiner();
            }
        }

        return merged.toArray(new String[0]);
    }

    private static boolean isField(@NotNull String line, @NotNull List<String> validKeys) {
        return line.contains(QuestionnaireExtraction.KEY_VALUE_SEPARATOR) && validKeys.stream().anyMatch(line::contains);
    }

    @NotNull
    private static StringJoiner newValueStringJoiner() {
        return new StringJoiner(QuestionnaireExtraction.VALUE_LIST_SEPARATOR_1);
    }

    @NotNull
    private static String clean(@NotNull String entryText) {
        String cleanText = entryText;
        for (String str : TERMS_TO_CLEAN) {
            cleanText = cleanText.replace(str, Strings.EMPTY);
        }
        return cleanText;
    }
}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/QuestionnaireReader.kt
package com.hartwig.actin.clinical.feed.emc.questionnaire

import com.google.common.collect.Lists
import java.util.StringJoiner

internal object QuestionnaireReader {

    val TERMS_TO_CLEAN =
        setOf("{", "}", "\\tab", "\\li0", "\\ri0", "\\sa0", "\\sb0", "\\u000ci0", "\\ql", "\\par", "\\u000c2", "\\ltrch", "\\fi0", "\\f2")

    fun read(entryText: String, validKeys: List<String>): Array<String> {
        return merge(clean(entryText, validKeys).split("\\n").dropLastWhile { it.isEmpty() }.toTypedArray(), validKeys)
    }

    private fun merge(lines: Array<String>, validKeys: List<String>): Array<String> {
        val merged: MutableList<String> = Lists.newArrayList()
        var curLine = newValueStringJoiner()
        for (i in lines.indices) {
            curLine.add(lines[i])
            if (lines[i].trim { it <= ' ' }.isEmpty() || i == lines.size - 1 || isField(
                    lines[i + 1],
                    validKeys
                ) || lines[i + 1].trim { it <= ' ' }.isEmpty()
            ) {
                merged.add(curLine.toString())
                curLine = newValueStringJoiner()
            }
        }
        return merged.toTypedArray()
    }

    private fun isField(line: String, validKeys: List<String>): Boolean {
        return line.contains(QuestionnaireExtraction.KEY_VALUE_SEPARATOR) && validKeys.any { line.contains(it) }
    }

    private fun newValueStringJoiner(): StringJoiner {
        return StringJoiner(QuestionnaireExtraction.VALUE_LIST_SEPARATOR_1)
    }

    private fun clean(entryText: String, validKeys: List<String>): String {
        val cleanedText = TERMS_TO_CLEAN.fold(entryText) { acc, term ->
            acc.replace(term, "")
        }
        val pattern = "(${validKeys.joinToString("|")}):\\s*((\\\\n){2,})".toRegex()
        return pattern.replace(cleanedText) { match ->
            "${match.groupValues[1]}:\\n"
        }
    }
}

